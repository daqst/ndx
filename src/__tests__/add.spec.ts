import { createIndex, addDocumentToIndex } from "..";

const tokenizer = (s: string) => s.split(" ");
const filter = (s: string) => s;

test("should invoke tokenizer", () => {
  let v = "";
  const t = (s: string) => (v = s, tokenizer(s));
  const idx = createIndex<number>(1);
  const doc = { id: 1, text: "a b c" };
  addDocumentToIndex(idx, [(d) => d.text], t, filter, doc.id, doc);
  expect(v).toBe("a b c");
});

test("should invoke filter for each term", () => {
  const v: string[] = [];
  const f = (s: string) => (v.push(s), s);
  const idx = createIndex<number>(1);
  const doc = { id: 1, text: "a b c" };
  addDocumentToIndex(idx, [(d) => d.text], tokenizer, f, doc.id, doc);
  expect(v).toEqual(["a", "b", "c"]);
});

test("one document with three terms", () => {
  const idx = createIndex<number>(1);
  const docs = [{ id: 1, text: "a b c" }];
  docs.forEach((doc) => {
    addDocumentToIndex(idx, [(d) => d.text], tokenizer, filter, doc.id, doc);
  });
  expect(idx).toMatchSnapshot();
});

test("two documents with different terms", () => {
  const idx = createIndex<number>(1);
  const docs = [
    { id: 1, text: "a b c" },
    { id: 2, text: "d e f" },
  ];
  docs.forEach((doc) => {
    addDocumentToIndex(idx, [(d) => d.text], tokenizer, filter, doc.id, doc);
  });
  expect(idx).toMatchSnapshot();
});

test("two documents with shared terms", () => {
  const idx = createIndex<number>(1);
  const docs = [
    { id: 1, text: "a b c" },
    { id: 2, text: "b c d" },
  ];
  docs.forEach((doc) => {
    addDocumentToIndex(idx, [(d) => d.text], tokenizer, filter, doc.id, doc);
  });
  expect(idx).toMatchSnapshot();
});

test("document with empty text", () => {
  const idx = createIndex<number>(1);
  const docs = [{ id: 1, text: "" }];
  docs.forEach((doc) => {
    addDocumentToIndex(idx, [(d) => d.text], tokenizer, filter, doc.id, doc);
  });
  expect(idx).toMatchSnapshot();
});

test("document with multiple fields", () => {
  const idx = createIndex<number>(2);
  const docs = [{ id: 1, text1: "a b c", text2: "c d e" }];
  docs.forEach((doc) => {
    addDocumentToIndex(idx, [(d) => d.text1, (d) => d.text2], tokenizer, filter, doc.id, doc);
  });
  expect(idx).toMatchSnapshot();
});

test("document with unavailable field", () => {
  const idx = createIndex<number>(2);
  const docs = [{ id: 1, text1: "a b c" }];
  docs.forEach((doc) => {
    addDocumentToIndex(idx, [(d) => d.text1, (d) => (d as any).text2], tokenizer, filter, doc.id, doc);
  });
  expect(idx).toMatchSnapshot();
});

test("document with object key", () => {
  const docs = [{ id: 1, text1: "a b c" }];
  const idx = createIndex<typeof docs[0]>(2);
  docs.forEach((doc) => {
    addDocumentToIndex(idx, [(d) => d.text1, (d) => (d as any).text2], tokenizer, filter, doc, doc);
  });
  expect(idx).toMatchSnapshot();
});
