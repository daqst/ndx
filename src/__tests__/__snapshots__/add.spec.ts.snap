// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`document with empty text 1`] = `
Object {
  "docs": Map {
    1 => Object {
      "fieldLengths": Array [
        0,
      ],
      "key": 1,
    },
  },
  "fields": Array [
    Object {
      "avg": 0,
      "sum": 0,
    },
  ],
  "root": Object {
    "charCode": 0,
    "firstChild": null,
    "firstDoc": null,
    "next": null,
  },
}
`;

exports[`document with multiple fields 1`] = `
Object {
  "docs": Map {
    1 => Object {
      "fieldLengths": Array [
        3,
        3,
      ],
      "key": 1,
    },
  },
  "fields": Array [
    Object {
      "avg": 3,
      "sum": 3,
    },
    Object {
      "avg": 3,
      "sum": 3,
    },
  ],
  "root": Object {
    "charCode": 0,
    "firstChild": Object {
      "charCode": 101,
      "firstChild": null,
      "firstDoc": Object {
        "details": Object {
          "fieldLengths": Array [
            3,
            3,
          ],
          "key": 1,
        },
        "next": null,
        "termFrequency": Array [
          0,
          1,
        ],
      },
      "next": Object {
        "charCode": 100,
        "firstChild": null,
        "firstDoc": Object {
          "details": Object {
            "fieldLengths": Array [
              3,
              3,
            ],
            "key": 1,
          },
          "next": null,
          "termFrequency": Array [
            0,
            1,
          ],
        },
        "next": Object {
          "charCode": 99,
          "firstChild": null,
          "firstDoc": Object {
            "details": Object {
              "fieldLengths": Array [
                3,
                3,
              ],
              "key": 1,
            },
            "next": null,
            "termFrequency": Array [
              1,
              1,
            ],
          },
          "next": Object {
            "charCode": 98,
            "firstChild": null,
            "firstDoc": Object {
              "details": Object {
                "fieldLengths": Array [
                  3,
                  3,
                ],
                "key": 1,
              },
              "next": null,
              "termFrequency": Array [
                1,
                0,
              ],
            },
            "next": Object {
              "charCode": 97,
              "firstChild": null,
              "firstDoc": Object {
                "details": Object {
                  "fieldLengths": Array [
                    3,
                    3,
                  ],
                  "key": 1,
                },
                "next": null,
                "termFrequency": Array [
                  1,
                  0,
                ],
              },
              "next": null,
            },
          },
        },
      },
    },
    "firstDoc": null,
    "next": null,
  },
}
`;

exports[`document with object key 1`] = `
Object {
  "docs": Map {
    Object {
      "id": 1,
      "text1": "a b c",
    } => Object {
      "fieldLengths": Array [
        3,
        0,
      ],
      "key": Object {
        "id": 1,
        "text1": "a b c",
      },
    },
  },
  "fields": Array [
    Object {
      "avg": 3,
      "sum": 3,
    },
    Object {
      "avg": 0,
      "sum": 0,
    },
  ],
  "root": Object {
    "charCode": 0,
    "firstChild": Object {
      "charCode": 99,
      "firstChild": null,
      "firstDoc": Object {
        "details": Object {
          "fieldLengths": Array [
            3,
            0,
          ],
          "key": Object {
            "id": 1,
            "text1": "a b c",
          },
        },
        "next": null,
        "termFrequency": Array [
          1,
          0,
        ],
      },
      "next": Object {
        "charCode": 98,
        "firstChild": null,
        "firstDoc": Object {
          "details": Object {
            "fieldLengths": Array [
              3,
              0,
            ],
            "key": Object {
              "id": 1,
              "text1": "a b c",
            },
          },
          "next": null,
          "termFrequency": Array [
            1,
            0,
          ],
        },
        "next": Object {
          "charCode": 97,
          "firstChild": null,
          "firstDoc": Object {
            "details": Object {
              "fieldLengths": Array [
                3,
                0,
              ],
              "key": Object {
                "id": 1,
                "text1": "a b c",
              },
            },
            "next": null,
            "termFrequency": Array [
              1,
              0,
            ],
          },
          "next": null,
        },
      },
    },
    "firstDoc": null,
    "next": null,
  },
}
`;

exports[`document with unavailable field 1`] = `
Object {
  "docs": Map {
    1 => Object {
      "fieldLengths": Array [
        3,
        0,
      ],
      "key": 1,
    },
  },
  "fields": Array [
    Object {
      "avg": 3,
      "sum": 3,
    },
    Object {
      "avg": 0,
      "sum": 0,
    },
  ],
  "root": Object {
    "charCode": 0,
    "firstChild": Object {
      "charCode": 99,
      "firstChild": null,
      "firstDoc": Object {
        "details": Object {
          "fieldLengths": Array [
            3,
            0,
          ],
          "key": 1,
        },
        "next": null,
        "termFrequency": Array [
          1,
          0,
        ],
      },
      "next": Object {
        "charCode": 98,
        "firstChild": null,
        "firstDoc": Object {
          "details": Object {
            "fieldLengths": Array [
              3,
              0,
            ],
            "key": 1,
          },
          "next": null,
          "termFrequency": Array [
            1,
            0,
          ],
        },
        "next": Object {
          "charCode": 97,
          "firstChild": null,
          "firstDoc": Object {
            "details": Object {
              "fieldLengths": Array [
                3,
                0,
              ],
              "key": 1,
            },
            "next": null,
            "termFrequency": Array [
              1,
              0,
            ],
          },
          "next": null,
        },
      },
    },
    "firstDoc": null,
    "next": null,
  },
}
`;

exports[`one document with three terms 1`] = `
Object {
  "docs": Map {
    1 => Object {
      "fieldLengths": Array [
        3,
      ],
      "key": 1,
    },
  },
  "fields": Array [
    Object {
      "avg": 3,
      "sum": 3,
    },
  ],
  "root": Object {
    "charCode": 0,
    "firstChild": Object {
      "charCode": 99,
      "firstChild": null,
      "firstDoc": Object {
        "details": Object {
          "fieldLengths": Array [
            3,
          ],
          "key": 1,
        },
        "next": null,
        "termFrequency": Array [
          1,
        ],
      },
      "next": Object {
        "charCode": 98,
        "firstChild": null,
        "firstDoc": Object {
          "details": Object {
            "fieldLengths": Array [
              3,
            ],
            "key": 1,
          },
          "next": null,
          "termFrequency": Array [
            1,
          ],
        },
        "next": Object {
          "charCode": 97,
          "firstChild": null,
          "firstDoc": Object {
            "details": Object {
              "fieldLengths": Array [
                3,
              ],
              "key": 1,
            },
            "next": null,
            "termFrequency": Array [
              1,
            ],
          },
          "next": null,
        },
      },
    },
    "firstDoc": null,
    "next": null,
  },
}
`;

exports[`two documents with different terms 1`] = `
Object {
  "docs": Map {
    1 => Object {
      "fieldLengths": Array [
        3,
      ],
      "key": 1,
    },
    2 => Object {
      "fieldLengths": Array [
        3,
      ],
      "key": 2,
    },
  },
  "fields": Array [
    Object {
      "avg": 3,
      "sum": 6,
    },
  ],
  "root": Object {
    "charCode": 0,
    "firstChild": Object {
      "charCode": 102,
      "firstChild": null,
      "firstDoc": Object {
        "details": Object {
          "fieldLengths": Array [
            3,
          ],
          "key": 2,
        },
        "next": null,
        "termFrequency": Array [
          1,
        ],
      },
      "next": Object {
        "charCode": 101,
        "firstChild": null,
        "firstDoc": Object {
          "details": Object {
            "fieldLengths": Array [
              3,
            ],
            "key": 2,
          },
          "next": null,
          "termFrequency": Array [
            1,
          ],
        },
        "next": Object {
          "charCode": 100,
          "firstChild": null,
          "firstDoc": Object {
            "details": Object {
              "fieldLengths": Array [
                3,
              ],
              "key": 2,
            },
            "next": null,
            "termFrequency": Array [
              1,
            ],
          },
          "next": Object {
            "charCode": 99,
            "firstChild": null,
            "firstDoc": Object {
              "details": Object {
                "fieldLengths": Array [
                  3,
                ],
                "key": 1,
              },
              "next": null,
              "termFrequency": Array [
                1,
              ],
            },
            "next": Object {
              "charCode": 98,
              "firstChild": null,
              "firstDoc": Object {
                "details": Object {
                  "fieldLengths": Array [
                    3,
                  ],
                  "key": 1,
                },
                "next": null,
                "termFrequency": Array [
                  1,
                ],
              },
              "next": Object {
                "charCode": 97,
                "firstChild": null,
                "firstDoc": Object {
                  "details": Object {
                    "fieldLengths": Array [
                      3,
                    ],
                    "key": 1,
                  },
                  "next": null,
                  "termFrequency": Array [
                    1,
                  ],
                },
                "next": null,
              },
            },
          },
        },
      },
    },
    "firstDoc": null,
    "next": null,
  },
}
`;

exports[`two documents with shared terms 1`] = `
Object {
  "docs": Map {
    1 => Object {
      "fieldLengths": Array [
        3,
      ],
      "key": 1,
    },
    2 => Object {
      "fieldLengths": Array [
        3,
      ],
      "key": 2,
    },
  },
  "fields": Array [
    Object {
      "avg": 3,
      "sum": 6,
    },
  ],
  "root": Object {
    "charCode": 0,
    "firstChild": Object {
      "charCode": 100,
      "firstChild": null,
      "firstDoc": Object {
        "details": Object {
          "fieldLengths": Array [
            3,
          ],
          "key": 2,
        },
        "next": null,
        "termFrequency": Array [
          1,
        ],
      },
      "next": Object {
        "charCode": 99,
        "firstChild": null,
        "firstDoc": Object {
          "details": Object {
            "fieldLengths": Array [
              3,
            ],
            "key": 2,
          },
          "next": Object {
            "details": Object {
              "fieldLengths": Array [
                3,
              ],
              "key": 1,
            },
            "next": null,
            "termFrequency": Array [
              1,
            ],
          },
          "termFrequency": Array [
            1,
          ],
        },
        "next": Object {
          "charCode": 98,
          "firstChild": null,
          "firstDoc": Object {
            "details": Object {
              "fieldLengths": Array [
                3,
              ],
              "key": 2,
            },
            "next": Object {
              "details": Object {
                "fieldLengths": Array [
                  3,
                ],
                "key": 1,
              },
              "next": null,
              "termFrequency": Array [
                1,
              ],
            },
            "termFrequency": Array [
              1,
            ],
          },
          "next": Object {
            "charCode": 97,
            "firstChild": null,
            "firstDoc": Object {
              "details": Object {
                "fieldLengths": Array [
                  3,
                ],
                "key": 1,
              },
              "next": null,
              "termFrequency": Array [
                1,
              ],
            },
            "next": null,
          },
        },
      },
    },
    "firstDoc": null,
    "next": null,
  },
}
`;
